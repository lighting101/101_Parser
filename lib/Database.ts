import * as mysql from "mysql";
import {mysqlConfig} from "../config";

export default class Database
{
    private connection:mysql.Connection;

    constructor() {
        this.connection = mysql.createConnection(mysqlConfig);
    }

    query(sql:string, params:Array<any> = []):Promise<any> {
        return new Promise((resolve, reject) => {
            const finalSql = mysql.format(sql, params);
            this.connection.query(finalSql, (err, rows) => {
                if (err) {
                    reject(err);
                } else {
                    resolve(rows);
                }
            });
        });
    }

    beginTransaction():Promise<void> {
        return new Promise(((resolve, reject) => {
            this.connection.beginTransaction(err => {
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            })
        }))
    }

    commit():Promise<void> {
        return new Promise(((resolve, reject) => {
            this.connection.commit(err => {
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            })
        }))
    }

    rollback():Promise<void> {
        return new Promise(((resolve, reject) => {
            this.connection.rollback(err => {
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            })
        }))
    }

    end():Promise<void> {
        return new Promise(((resolve, reject) => {
            this.connection.end(err => {
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            })
        }))
    }
}