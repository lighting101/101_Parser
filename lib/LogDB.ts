import Database from "./Database";
import ILog from "./Interfaces/ILog";

export enum ErrorLevels {
    debug,
    info,
    error
}

type LogDBConfig = {
    Database: ErrorLevels,
    Console: ErrorLevels
}

const defaultConfig:LogDBConfig = {
    Database: ErrorLevels.info,
    Console: ErrorLevels.error
}

export default class LogDB implements ILog {
    private config:LogDBConfig;
    private db:Database;
    private moduleName:string;
    protected errorLevelText = ['debug', 'info', 'error'];

    constructor(moduleName = '', config:LogDBConfig = defaultConfig) {
        this.db = new Database();
        this.config = config;
        this.moduleName = moduleName;
    }

    async info(msg: string): Promise<void> {
        const level = ErrorLevels.info;
        await this.drawMessage(msg, level);
    }

    async debug(msg: string): Promise<void> {
        const level = ErrorLevels.debug;
        await this.drawMessage(msg, level);
    }

    async error(msg: string): Promise<void> {
        const level = ErrorLevels.error
        await this.drawMessage(msg, level);
    }

    protected level2Text(level:number):string {
        return this.errorLevelText[level];
    }

    private async drawMessage(message:string, level:number) {
        if (level >= this.config.Console) {
            console.log(message);
        }

        if (level >= this.config.Database) {
            const sql = 'insert into `log` set ?'
            await this.db.query(sql, [{ message, level: this.level2Text(level) }])
        }
    }
}