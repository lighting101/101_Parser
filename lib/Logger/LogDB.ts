import Database from "../Database";
import LogBase from "./LogBase";

const db = new Database();

export default class LogDB extends LogBase{
    protected errorLevelText = ['debug', 'info', 'error'];

    protected level2Text(level:number):string {
        return this.errorLevelText[level];
    }

    protected async drawMessage(message:string, level:number):Promise<string> {
        const formatMessage = await super.drawMessage(message, level);

        const sql = 'insert into `log` set ?'
        await db.query(sql, [{ message: formatMessage, level: this.level2Text(level) }])

        return formatMessage;
    }
}