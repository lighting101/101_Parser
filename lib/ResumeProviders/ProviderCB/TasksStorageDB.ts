import {TaskFormat} from "../../../common";
import Database from "../../Database";
import {TASKS_LIMIT, maxHoursTaskCanProcessing} from "../../../config";
import ITasksStorage from "./Interfaces/ITaskStorage";
import {TaskType} from "../../../common";
import ILog from "../../Interfaces/ILog";
import LogFactory from "../../LogFactory";

type TasksDB = {
    id: number,
    type: string,
    data: string,
    controlhash: string
};

export default class TasksStorageDB implements ITasksStorage
{
    private log:ILog;
    private db:Database;

    constructor(logProvider = LogFactory('TaskStorageDB'),
                oDB = new Database()) {

        this.log = logProvider;
        this.db = oDB;
    }

    async markDone(task: TaskFormat): Promise<void> {
        const sql = 'update `tasks` set `done` = 1 where `id` = ?';
        await this.db.query(sql, [ task.id ]);
    }

    async getTasks(): Promise<TaskFormat[]> {
        await this.log.debug(`getTasks() started`);

        let tasks:TasksDB[];

        await this.log.debug(`getTasks() starting transaction...`);
        await this.db.beginTransaction();
        await this.log.debug(`getTasks() transaction has been started`);

        // get tasks with status = 1 (waiting for process)
        const sql1 = 'SELECT `id`, `type`, `data` ' +
            'FROM `tasks` ' +
            'WHERE `done` = 0 and `processing` < date_sub(now(), interval ? hour) ' +
            'order by `type` asc, `ts` asc ' +
            'limit ?';

        try {
            await this.log.debug(`getTasks() trying to get tasks from the DB...`);
            tasks = await this.db.query(sql1, [ maxHoursTaskCanProcessing, TASKS_LIMIT ]);
            await this.log.debug(`getTasks() got ${tasks.length} rows from the DB`);
        } catch (e) {
            await this.log.debug(`getTasks() caught error ${e.name}: ${e.message}`);
            await this.log.debug(`getTasks() starting rollback operation...`);
            await this.db.rollback();
            await this.log.debug(`getTasks() rollback done`);
            throw e;
        }

        if (tasks.length === 0) {
            throw new Error('No any tasks in DB');
        }

        const sql2 = 'update `tasks` set `processing` = now() where `id` in (?)';

        const taskIDs = tasks.map(task => task.id)
        await this.log.debug(`getTasks() prepare query to update the 'processing' field (amount of IDs=${taskIDs.length})`);
        await this.db.query(sql2, [ taskIDs ]);
        await this.log.debug(`getTasks() query was done`);

        await this.log.debug(`getTasks() committing transaction...`);
        await this.db.commit();
        await this.log.debug(`getTasks() committed`);

        const returnTasks:TaskFormat[] = [];

        for (const task of tasks) {
            const type = <TaskType> task.type;

            const data = <{
                kind?: 'branch'|'resume',
                keywords?: string,
                city?: string,
                state?: string,
                freshness?: number,
                page?: number,
                rowsPerPage?: number,
                resumeID?: string
            }> JSON.parse(task.data);

            const id = <number> task.id;

            // Does have a "kind" field in the data?
            if (!('kind' in data)) {
                if (!(type === 'branch' || type === 'resume')) {
                    const errMsg = 'getTasks() got undefined task type';
                    await this.log.error(errMsg);
                    throw new Error(errMsg);
                }

                data.kind = type;
            }

            // Checking to have other needing fields in the data to according with the Task format
            if ( !(('resumeID' in data && data.kind === 'resume') ||
                ('keywords' in data &&
                    'city' in data &&
                    'state' in data &&
                    'freshness' in data &&
                    'page' in data &&
                    'rowsPerPage' in data &&
                    data.kind === 'branch'))) {

                const errMsg = `getTasks() Error in according a data object with the Task format: ${JSON.stringify(data)}`;
                await this.log.error(errMsg);
                throw new Error(errMsg);
            }

            returnTasks.push( <TaskFormat> {data, id} );
        }

        await this.log.debug(`getTasks() finish, found ${returnTasks.length} tasks`);

        return returnTasks;
    }

    async putTask(task: TaskFormat): Promise<void> {
        const type = task.data.kind;

        if (!(type === 'resume' || type === 'branch')) {
            const errorMsg = 'Undefined kind field in the task';
            await this.log.error(errorMsg);
            throw new Error(errorMsg);
        }

        const sql = 'insert ignore into `tasks` set type=?, data=?, controlhash=md5(data)';
        await this.db.query(sql, [ type, JSON.stringify(task.data) ]);
    }

    async putTasks(tasks: Array<TaskFormat>): Promise<void> {
        for (const task of tasks) {
            await this.putTask(task);
        }
    }
}