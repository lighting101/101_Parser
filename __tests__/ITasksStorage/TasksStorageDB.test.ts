import TasksStorageDB from "../../lib/ResumeProviders/ProviderCB/TasksStorageDB";
import LogDB from "../../lib/LogDB";
import Database from "../../lib/Database";

jest.mock("../../lib/LogDB");
jest.mock("../../lib/Database");

beforeEach(() => {
    jest.resetAllMocks();
});

function classFactory() {
    const storage = new Database();

    const ts = new TasksStorageDB(
        new LogDB('test'),
        storage);

    return {ts, storage};
}

describe('getTasks()', () => {
    it('Amount of the tasks must be equivalent to amount of the results', async () => {
        expect.assertions(1)

        const {ts, storage} = classFactory();

        const testTasks = [
            {id: 1, type: 'resume', data: '{"resumeID": "R0000000"}'},
            {id: 2, type: 'resume', data: '{"resumeID": "R0000001"}'},
            {id: 3, type: 'resume', data: '{"resumeID": "R0000002"}'},
            {id: 4, type: 'resume', data: '{"resumeID": "R0000003"}'},
            {id: 5, type: 'resume', data: '{"resumeID": "R0000004"}'},
        ]

        storage.query = jest.fn().mockResolvedValue(testTasks);

        const result = await ts.getTasks();

        expect(result.length).toBe(testTasks.length);
    })

    it('Accordance of the well-result snapshot', async () => {
        expect.assertions(1)

        const {ts, storage} = classFactory();

        const testTasks = [
            {id: 1, type: 'resume', data: '{"resumeID": "R0000000"}'},
            {id: 2, type: 'resume', data: '{"resumeID": "R0000001"}'},
            {id: 3, type: 'resume', data: '{"resumeID": "R0000002"}'},
            {id: 4, type: 'resume', data: '{"resumeID": "R0000003"}'},
            {id: 5, type: 'resume', data: '{"resumeID": "R0000004"}'},
        ]

        storage.query = jest.fn().mockResolvedValue(testTasks);

        const result = await ts.getTasks();

        expect(result).toMatchSnapshot();
    })

    it('If dont have any tasks in DB the method throws a specific exception', async () => {
        expect.assertions(1)

        const {ts, storage} = classFactory();

        storage.query = jest.fn().mockResolvedValue([]);

        try {
            await ts.getTasks();
        } catch (e) {
            expect(e.message).toContain('No any tasks in DB')
        }
    })
})
