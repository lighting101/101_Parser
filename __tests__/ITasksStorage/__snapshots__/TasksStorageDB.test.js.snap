// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTasks() Accordance of the well-result snapshot 1`] = `
Array [
  Object {
    "data": Object {
      "kind": "resume",
      "resumeID": "R0000000",
    },
    "id": 1,
  },
  Object {
    "data": Object {
      "kind": "resume",
      "resumeID": "R0000001",
    },
    "id": 2,
  },
  Object {
    "data": Object {
      "kind": "resume",
      "resumeID": "R0000002",
    },
    "id": 3,
  },
  Object {
    "data": Object {
      "kind": "resume",
      "resumeID": "R0000003",
    },
    "id": 4,
  },
  Object {
    "data": Object {
      "kind": "resume",
      "resumeID": "R0000004",
    },
    "id": 5,
  },
]
`;
